### PR TITLE
better handling of sudo flag replacement (#76389)

### PR DESCRIPTION
**Pull request by bcoca**
* better handling of sudo flag replacement

(cherry picked from commit b02168d644dfaff1f76f52931d7b41485c2753a3)

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

become/sudo

### CODE REVIEW

File: lib/ansible/plugins/become/sudo.py [hunk 1]
```diff
@@ -94,7 +96,16 @@ def build_become_command(self, cmd, shell):
...
-                flags = flags.replace('-n', '')
+                reflag = []
+                for flag in shlex.split(flags):
+                    if flag in ('-n', '--non-interactive'):
+                        continue
+                    elif not flag.startswith('--'):
+                        # handle -XnxxX flags only
+                        flag = re.sub(r'^(-\w*)n(\w*.*)', r'\1\2', flag)
+                    reflag.append(flag)
+                flags = shlex.join(reflag)
```
Comment (mattclay) [inline lib/ansible/plugins/become/sudo.py]: `shlex.join` was added in Python 3.8, so an alternative is needed here.

File: lib/ansible/plugins/become/sudo.py [hunk 2]
```diff
@@ -94,7 +96,20 @@ def build_become_command(self, cmd, shell):
...
+                        continue
+                    elif not flag.startswith('--'):
+                        # handle -XnxxX flags only
+                        flag = re.sub(r'^(-\w*)n(\w*.*)', r'\1\2', flag)
+                    reflag.append(flag)
+                try:
+                    flags = shlex.join(reflag)
+                except AttributeError:
+                    # for older pythons
+                    flags = ' '.join([shlex.quote(a) for a in reflag])
```
Comment (mattclay) [inline lib/ansible/plugins/become/sudo.py]: `shlex.quote` isn't available on Python 2.x.

Comment (bcoca) [reply lib/ansible/plugins/become/sudo.py]: i could swear we were using that one for a long time ... I'll update

File: lib/ansible/plugins/become/sudo.py [hunk 3]
```diff
@@ -94,7 +97,20 @@ def build_become_command(self, cmd, shell):
...
             self.prompt = '[sudo via ansible, key=%s] password:' % self._id
             if flags:  # this could be simplified, but kept as is for now for backwards string matching
-                flags = flags.replace('-n', '')
+                reflag = []
+                for flag in shlex.split(flags):
+                    if flag in ('-n', '--non-interactive'):
+                        continue
+                    elif not flag.startswith('--'):
+                        # handle -XnxxX flags only
+                        flag = re.sub(r'^(-\w*)n(\w*.*)', r'\1\2', flag)
```
Comment (sivel) [inline lib/ansible/plugins/become/sudo.py]: There appears to be a problem here on Py2.6 in that it's turning the options into:

```
-\x00\x00\x00H\x00\x00\x00
```

Example of `reflag`:

```
['-\x00\x00\x00H\x00\x00\x00', '\x00\x00\x00-\x00\x00\x00S\x00\x00\x00', '\x00\x00\x00-\x00\x00\x00n\x00\x00\x00']
```

Comment (bcoca) [reply lib/ansible/plugins/become/sudo.py]: pushed flags=re.U to fix

Comment (mattclay) [reply lib/ansible/plugins/become/sudo.py]: Python 2.6's `re.sub` doesn't accept `flags` -- but `re.compile` does.

Comment (bcoca) [reply lib/ansible/plugins/become/sudo.py]: had just figured that out in ssh_debug backport, thanks!

**Comment by sivel** [comment]
I'm going to close this out since 2.11 will move to security only on the 23rd, and this PR hasn't passed tests.

**Pull request closed without merging**